Tests/JetworkingTests/Mocks/MockExecuter.swift

```diff
@@ -0,0 +1,106 @@
+import Foundation
+@testable import Jetworking
+
+struct MockCancellable: CancellableRequest {
+    var identifier: Int = Int.random(in: (Int.min..<Int.max))
+    var onCancelCalled: ((Int) -> Void)?
+
+    init(onCancelCalled: ((Int) -> Void)?) {
+        self.onCancelCalled = onCancelCalled
+    }
+
+    func cancel() {
+        onCancelCalled?(identifier)
+    }
+}
+
+final class MockExecuter: RequestExecutor {
+    let session: URLSession
+    let encoder: JSONEncoder = JSONEncoder()
+    let defaultCompletionDelay: TimeInterval = 0.5
+    let responseCode: Int = 200
+    let headerFields: [String: String]? = nil
+    var isCancelled: Bool = false
+
+    static var completionDelayForRequest: ((URLRequest) -> TimeInterval)?
+
+    init(session: URLSession) {
+        self.session = session
+    }
+
+    func send(
+        request: URLRequest,
+        _ completion: @escaping ((Data?, URLResponse?, Error?) -> Void)
+    ) -> CancellableRequest? {
+        let completionDelay = MockExecuter.completionDelayForRequest?(request) ?? defaultCompletionDelay
+        DispatchQueue.main.asyncAfter(deadline: .now() + completionDelay) { [weak self] in
+            guard self?.isCancelled == false  else { return completion(nil, nil, URLError(.cancelled)) }
+
+            self?.execute(
+                request: request,
+                completion: completion
+            )
+        }
+
+        return MockCancellable { [weak self] _ in
+            self?.isCancelled = true
+        }
+    }
+
+    private func execute(
+        request: URLRequest,
+        completion: @escaping ((Data?, URLResponse?, Error?) -> Void)
+    ) {
+        guard
+            let url = request.url,
+            let method = request.httpMethod
+        else {
+            return
+        }
+
+        switch (url.absoluteString, HTTPMethod(rawValue: method.uppercased())) {
+        case ("https://www.jamitlabs.com/somePath", .GET),
+             ("https://www.jamitlabs.com/somePath", .DELETE),
+             ("https://www.jamitlabs.com/somePath", .PATCH),
+             ("https://www.jamitlabs.com/somePath", .PUT),
+             ("https://www.jamitlabs.com/somePath", .POST):
+            completion(
+                try? encoder.encode(MockBody(foo1: "SomeFoo", foo2: "AnotherFoo")),
+                request.toHTTPURLResponse(
+                    with: responseCode,
+                    andHeaderFields: headerFields
+                ),
+                nil
+            )
+
+        case ("https://www.jamitlabs.com/somePath?SomeKey=SomeValue", .GET):
+            completion(
+                try? encoder.encode(MockBody(foo1: "SomeFoo", foo2: "AnotherFoo")),
+                request.toHTTPURLResponse(
+                    with: responseCode,
+                    andHeaderFields: headerFields
+                ),
+                nil
+            )
+
+        default:
+            completion(nil, nil, URLError(.cancelled))
+        }
+    }
+}
+
+private extension URLRequest {
```

Comment:
Can we extract this? It is somehow hidden within this executer and I nearly oversaw it. Might be confusing for others as well.

Reply:
It's a private extension since it is just used in this file/context. If I extract this into a separate file I've to adjust the visibility.

Reply:
Okay since it is only used within the tests this extension can stay here.